File: tags/javachinesepinyin-1.3/src/main/java/pinyin/WordToPinyin.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package pinyin;

import pinyin.util.Pair;
import pinyin.util.Window;
import hmm.*;
import hmm.ngram.*;
import java.io.*;
import java.util.*;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

/**
 *
 * @author ray
 */
public class WordToPinyin {

    Viterbi<String, Character> viterbi = new Viterbi<String, Character>();
    //
    Map<Integer, Integer> pii = new HashMap<Integer, Integer>();
    Map<Integer, Map<Integer, Integer>> emisMatrix = new HashMap<Integer, Map<Integer, Integer>>();
    NodeBank<String, Node<String>> stateBank = new NodeBank<String, Node<String>>();
    NodeBank<Character, Node<Character>> observeBank = new NodeBank<Character, Node<Character>>();
    //
    TreeNode<String> ngram = new TreeNode<String>();
    TreeNodeSortor<String> sortor = null;
    Comparator<String> comparator = null;

    public WordToPinyin() {
        sortor = new TreeNodeBinarySort<String>();
        comparator = new Comparator<String>() {

            @Override
            public int compare(String t, String t1) {
                return t.compareTo(t1);
            }
        };

        sortor.setComparator(new Comparator<TreeNode<String>>() {

            @Override
            public int compare(TreeNode<String> t, TreeNode<String> t1) {
                return t.getKey().compareTo(t1.getKey());
            }
        });
    }

    public NodeBank<Character, Node<Character>> getObserveBank() {
        return observeBank;
    }

    public NodeBank<String, Node<String>> getStateBank() {
        return stateBank;
    }

    public void setN(int n) {
        viterbi.setN(n);
    }

    public void init() {

//        ngram.buildIndex(ngram.getCount(), sortor);

        viterbi.setObserveBank(observeBank);
        viterbi.setStateBank(stateBank);

        Transition<String> tran = new Transition<String>(ngram, stateBank);
        tran.setComparator(comparator);
        tran.setSortor(sortor);

        Emission emis = new Emission(emisMatrix);
        emis.setObserveBank(observeBank);

        Pi pi = new Pi(pii);

        viterbi.setPi(pi);
        viterbi.setTran(tran);
        viterbi.setE(emis);

        pii.clear();
//        tranMatrix.clear();
        emisMatrix.clear();
        pii = null;
//        tranMatrix = null;
        emisMatrix = null;
    }

    public List<Node<String>> classify(Character[] o) throws ObserveListException {
        List<Character> observeList = new ArrayList<Character>(o.length);
        observeList.addAll(Arrays.asList(o));
        return classify(observeList);
    }

    public List<Node<String>> classify(List<Character> o) throws ObserveListException {
        return viterbi.caculateWithLog(o);
    }

    public HmmResult viterbi(char[] o) throws ObserveListException {
        List<Character> observeList = new ArrayList<Character>(o.length);
        for (Character ob : o) {
//            Node observe = observeBank.get(ob);
            observeList.add(ob);
        }
        return viterbi(observeList);
    }

    public HmmResult viterbi(List<Character> o) throws ObserveListException {
        return viterbi.caculateHmmResult(o);
    }

    public void save(String filename) {
        try {
            ngram.buildIndex(ngram.getCount(), sortor);
//            ngram.cutCountLowerThan(Flag.getInstance().getNgramMinimumShowTimes());
            ObjectOutputStream oos = new ObjectOutputStream(new GZIPOutputStream(new FileOutputStream(filename)));
            oos.writeObject(stateBank);
            oos.writeObject(observeBank);
            oos.writeObject(pii);
//            oos.writeObject(tranMatrix);
            oos.writeObject(emisMatrix);
            oos.writeObject(ngram);
            oos.close();
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    public void load(String filename) {
        File file = new File(filename);
        InputStream is = null;
        if (!file.exists()) {
            is = this.getClass().getClassLoader().getResourceAsStream("ptw.m");
            load(is);
        } else {
            try {
                is = new FileInputStream(file);
                load(is);
            } catch (FileNotFoundException ex) {
                ex.printStackTrace();
            }
        }
    }

    public void load(InputStream is) {
        try {
            ObjectInputStream ois = new ObjectInputStream(new GZIPInputStream(is));
            stateBank = (NodeBank<String, Node<String>>) ois.readObject();
            observeBank = (NodeBank<Character, Node<Character>>) ois.readObject();
            pii = (Map<Integer, Integer>) ois.readObject();
//            tranMatrix = (Map<String, Map<Integer, Integer>>) ois.readObject();
            emisMatrix = (Map<Integer, Map<Integer, Integer>>) ois.readObject();
            ngram = (TreeNode<String>) ois.readObject();
            is.close();
        } catch (ClassNotFoundException ex) {
            ex.printStackTrace();
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    public void train(String filename) {
        File file = new File(filename);
        if (file.exists()) {
            System.out.println("[WordToPinyin] train file " + file.getAbsolutePath());
            try {
                BufferedReader br = new BufferedReader(
                        new InputStreamReader(
                        new FileInputStream(file), "UTF-8"));
                String line = br.readLine();
                Window<Pair<String, String>> window = new Window<Pair<String, String>>(Flag.getInstance().n);
                Pair<String, String> cur = null;
                while (null != line) {
                    cur = parseLine(line.trim());

                    if (null == cur) {
                        window.clear();
                    } else {
                        window.add(cur);
                        String o = cur.getKey();
                        boolean isHeadOfWord = o.startsWith("Head");
                        o = isHeadOfWord ? o.substring(4) : o;
                        Node<Character> observe = new Observe<Character>(Character.valueOf(o.charAt(0)));
                        observe = observeBank.add(observe);
                        String s = cur.getValue();
                        Node<String> state = new State(s);
                        state = stateBank.add(state);

                        //Pii
//                        if (isHeadOfWord)
                        {
                            int index = state.getIndex();
                            int c = pii.containsKey(index) ? pii.get(index) + 1 : 1;
                            pii.put(index, c);
                        }

                        //Transition
                        Pair<String, String> array[] = window.toArray(new Pair[0]);
//                        for (int i = 0; i < array.length; i++) {
                        statisticNGram(array);
//                        }

                        //Emission
                        int si = state.getIndex();
                        int oi = observe.getIndex();
                        Map<Integer, Integer> row = null;
                        if (emisMatrix.containsKey(si)) {
                            row = emisMatrix.get(si);
                        } else {
                            row = new HashMap<Integer, Integer>();
                            emisMatrix.put(si, row);
                        }
                        int count = row.containsKey(oi) ? row.get(oi) + 1 : 1;
                        row.put(oi, count);
                    }
                    line = br.readLine();
                }
                br.close();
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
    }

    private void statisticNGram(Pair<String, String> array[]) {
        String[] ch = new String[array.length];
        for (int i = 0; i < ch.length; i++) {
            ch[i] = array[i].getValue();
        }

        ngram.insert(ch, sortor, comparator);
    }

    private Pair<String, String> parseLine(String line) {
        Pair<String, String> ret = null;
        if ("".equals(line) || line.startsWith("c")) {
            return ret;
        }

        String[] pair = line.split("\\s");

        if (null != pair && pair.length == 2) {
            ret = new Pair<String, String>(pair[0], pair[1]);
        }

        return ret;
    }

    public static void main(String[] args) {
        if (args.length > 0) {
            WordToPinyin c = new WordToPinyin();
            File dir = new File(args[0]);
            if (dir.isDirectory() && dir.exists()) {
                for (File f : dir.listFiles()) {
                    String fname = f.getName();
                    if (fname.endsWith(".txt")) {
                        c.train(f.getAbsolutePath());
                    }
                }
                c.save("wtp.m");
            }
        } else {
            System.out.println("Please input directory of corpus:\n\t/>./train.sh ../corpus/LCMC");
        }
    }
}

File: lib.classifier/src/main/scala/pinyin/classifier/WordToPinyin.scala
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package pinyin.classifier

import hmm._
import ngram._
import pinyin.utils._

import java.io._
import java.util._
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

import scala.collection.JavaConversions._
import scala.collection.mutable.Map

class WordToPinyin extends HMMClassifier[String, Character] {
  var pii = Map[Int, Int]()
  var emisMatrix = Map[Int, Map[Int, Int]]()
  var stateBank = new NodeBank[String]
  var observeBank = new NodeBank[Character]();
  var ngram: TreeNode[String] = new TreeNodeImpl[String]("r");

  override def init() {
    ngram.buildIndex(ngram.getCount)

    viterbi.stateBank = stateBank
    viterbi.observeBank = observeBank

    viterbi.tran = Transition[String](ngram, stateBank)

    viterbi.e = Emission(emisMatrix)
    viterbi.e.observeBank = observeBank

    viterbi.pi = Pi(pii)

    pii.clear
    emisMatrix.clear
    pii = null
    emisMatrix = null
  }

  override def getStateBy(id: Int): String = stateBank.get(id).name

  override def save(filename: String): Unit = {
    try {
      ngram.buildIndex(ngram.getCount);
//            ngram.cutCountLowerThan(Flag.getInstance().getNgramMinimumShowTimes());
      val oos = new ObjectOutputStream(new GZIPOutputStream(new FileOutputStream(filename)));
      oos.writeObject(stateBank);
      oos.writeObject(observeBank);
      oos.writeObject(pii);
      oos.writeObject(emisMatrix);
      oos.writeObject(ngram);
      oos.close();
    } catch {
      case ex =>
        ex.printStackTrace();
    }
  }

  override def load(is: InputStream): Unit = {
    try {
      val ois = new ObjectInputStream(new GZIPInputStream(is));
      stateBank = ois.readObject().asInstanceOf[NodeBank[String]];
      observeBank = ois.readObject().asInstanceOf[NodeBank[Character]];
      pii = ois.readObject().asInstanceOf[Map[Int,Int]];
      emisMatrix = ois.readObject().asInstanceOf[Map[Int,Map[Int,Int]]];
      ngram = ois.readObject().asInstanceOf[TreeNode[String]];
      is.close();
    } catch {
      case ex =>
        ex.printStackTrace();
    }
  }

  def parseLine(line: String, reverse: Boolean):(String,String) = {
    if ("".equals(line) || line.startsWith("c")) {
      null
    } else {
      val pair = line.split("\\s");
      if (null != pair && pair.length == 2) {
        if(!reverse){
          (pair(0), pair(1))
        }else{
          (pair(1),pair(0))
        }
      } else {
        null
      }
    }
  }

  override def train(filename: String): Unit = {
    if(filename.endsWith(".txt")){
      trainHMM(filename)
    }
  }

  private def statisticNGram(list: List[(String,String)]): Unit = {
    val chs = new Array[String](list.size())
    for {
      i <- 0 until list.size()
      val t = list.get(i)
      val c = t._1
    } {
      chs(i) = c
    }
    ngram.insert(chs);
  }

  def trainHMM(filename: String): Unit = {
    val file = new File(filename)
    if (file.exists()) {
      println("[WordToPinyin] train file " + file.getAbsolutePath())
      try {
        val br = new BufferedReader(
          new InputStreamReader(
            new FileInputStream(file), "UTF-8"))
        var line = br.readLine
        val window = new Window[(String,String)](Flag.n)
        while (null != line) {
          val cur = parseLine(line.trim, true)

          if (null == cur) {
            window.clear()
          } else {
            window.add(cur)
            val s = cur._1//if(isHeadOfWord) cur._1.substring(4) else cur._1
            val state = stateBank.add(s)
            val isHeadOfWord = cur._2.startsWith("Head")
            val o = Character.valueOf(if(isHeadOfWord) cur._2.substring(4).charAt(0) else cur._2.charAt(0))
            val observe = observeBank.add(o)

            //Pii
//            if(isHeadOfWord)
            {
              val index = state.index
              val c = pii.getOrElse(index,0) + 1
              pii.put(index, c)
            }

            //Transition
            val array = window.toList
            statisticNGram(array)

            //Emission
            val si = state.index
            val oi = observe.index
            val row = emisMatrix.getOrElseUpdate(si, Map[Int,Int]())
            val count = row.getOrElse(oi, 0) + 1
            row.put(oi, count)
          }
          line = br.readLine;
        }
        br.close();
      } catch {
        case ex: IOException =>
          ex.printStackTrace();
      }
    }
  }

}
